libc/bsearch.c

```diff
@@ -0,0 +1,213 @@
+
+static struct node table[TABSIZE];
+
+static void prepare(char *names_arg[])
+{
+	int i;
+
+	for (i = 0; i < TABSIZE; i++) {
+		if ((table[i].string = (char *)malloc(strlen(names_arg[i]) + 1)) == NULL) {
+			TEST_MESSAGE("malloc failed");
+			exit(-1);
+		}
+		if (strcpy(table[i].string, names_arg[i]) != table[i].string) {
+			TEST_MESSAGE("strcpy failed");
+			exit(-1);
+		}
+		table[i].length = strlen(names_arg[i]);
+	}
+}
+
+
+/* bsearch can't change content of table, but may reorder elements */
+static int check_tab(char *names_arg[])
+{
+	int i, j, found;
+	for (i = 0; i < TABSIZE; i++) {
+		found = 0;
+		for (j = 0; j < TABSIZE; j++) {
+			if (strcmp(names_arg[i], table[j].string) == 0 && table[i].length == table[j].length) {
+				found = 1;
+				break;
+			}
+		}
+		if (found == 0) {
+			return found;
+		}
+	}
+
+	return found;
+}
+
+
+/* compare function used by bsearch */
+static int node_cmp(const void *node1, const void *node2)
+{
+	return strcmp(((struct node *)node1)->string, ((struct node *)node2)->string);
+}
+
+
+TEST_GROUP(bsearch);
+
+
+TEST_SETUP(bsearch)
+{
+}
+
+TEST_TEAR_DOWN(bsearch)
+{
+}
+
+
+TEST(bsearch, zero_el)
+{
+	struct node *rval, node;
+	node.string = "adam";
+	node.length = strlen(node.string);
+
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		0, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+}
+
+
+TEST(bsearch, find_node_in)
+{
+	int i;
+	struct node *rval, node;
+
+	/* check for even number of elements */
+	for (i = 0; i < TABSIZE; i++) {
+		node.string = names[i];
+		node.length = strlen(names[i]);
+		rval = (struct node *)bsearch((void *)&node, (void *)table,
+			(size_t)TABSIZE, sizeof(struct node), node_cmp);
+		TEST_ASSERT_NOT_NULL(rval);
+		TEST_ASSERT_EQUAL_MEMORY(&table[i], rval, sizeof(*rval));
+	}
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+
+	/* check for odd number of elements */
+	for (i = 0; i < TABSIZE - 1; i++) {
+		node.string = names[i];
+		node.length = strlen(names[i]);
+		rval = (struct node *)bsearch((void *)&node, (void *)table,
+			(size_t)TABSIZE - 1, sizeof(struct node), node_cmp);
+		TEST_ASSERT_NOT_NULL(rval);
+		TEST_ASSERT_EQUAL_MEMORY(&table[i], rval, sizeof(*rval));
+	}
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+}
+
+
+TEST(bsearch, find_node_not_in)
+{
+	struct node *rval, node;
+	char name[8];
+
+	TEST_ASSERT_EQUAL_MEMORY(name, strcpy(name, "phoenix"), sizeof(name));
+	node.string = name;
+	node.length = strlen(name);
+	/* check for even number of elements */
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		(size_t)TABSIZE, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+
+	/* check for odd number of elements */
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		(size_t)TABSIZE - 1, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+}
+
+
+TEST(bsearch, not_sorted)
```

Comment:
We could also look for all the elements there.

Reply:
To do this i would need to prepare test_table per every elements which is unnecessary effort in my opinion.